manual/manual/tutorials/moduleexamples.etex

```diff
@@ -178,14 +178,17 @@ module type PRIOQUEUE_WITH_OPT =
 \section{Functors}
 \pdfsection{Functors}
 
-Functors are ``functions'' from structures to structures. They are used to
-express parameterized structures: a structure \var{A} parameterized by a
+Functors let you create parameterized signatures and then provide the
```

Comment:
I was going on @djc's comments in Mantis.

My comments in Mantis:

```
Page 31:

“Functors are ‘functions’ from structures to structures” They are not general functions at
all, are they? As far as I can tell, they only allow substitution for specific parameters.
So the first sentence is misleading and would best be removed. The paragraph could begin
with “Functors let you create parameterized signatures and then provide the parameter(s)
to get a specific  implementation. For instance …”
```

@djc's comment in Mantis:

```
- Page 31: one sentence contradicts the other, functors are "'functions' from structures to
 structures" and hence not general at all. They are however functions in the sence that they
can be partially applied, with all the fun which comes with that. The module system is much
more powerful than your reduced description implies, but it would be worth rewording it to
 reflect both the simpler interpretation and their expressive power.
```

The first paragraph of Chapter 1 promises "no prior exposure to functional languages is required."  The idea of having a function from a structure to a structure or from a module to a module is not something at all familiar to someone used to conventional languages.  It is also the first sentence of a new chapter, and thus baffling.  Therefore I put in text that describes what you can do with functors, which seems more understandable.  What do you think?

Reply:
Pretty sure you didn't mean to invoke me, here. Maybe this dra person?

Reply:
Yes, sorry, I will fix.

Reply:
> It is also the first sentence of a new chapter, and thus baffling.

I think I see your point, starting with the right definition and explaining it might be an academic idiosyncrasy here. An option would be to keep the first sentence or amend it slightly and then insert a simpler example right after to give readers something concrete to focus on.

Reply:
How about the following (replacing both paragraphs)?

```
Functors are specialized ``functions'' from modules to modules. Functors
let you create parameterized module signatures and then provide the parameter(s)
to get a specific implementation.  For instance, a "Set" module implementing sets
as sorted lists could be parameterized to work with any module (such as "OrderedString")
that provides a type and a comparison method "compare"):
```

Reply:
I like the general idea, but I have few minor remarks:

> Functors are specialized ``functions'' from modules to modules.

I am not sure about the `specialized` adjective, it may lead the reader to think that there is a more generic construction available somewhere in the language which is not the case.

> Functors let you create parameterized module signatures

I would advocate to remove `signatures` to make clearer the distinction between module and module types: `Functors let you create parameterized modules`

> with any module (such as "OrderedString")

I am wondering if it would be better to move `(such as "OrderedString")` after the characterisation of the module:  `with any module that provides a type and a comparison method "compare"  (such as "OrderedString")`

Reply:
Good suggestions.  I also changed "provide the parameters(s)" to "provide other modules as parameter(s)" to be a bit more explicit.  It now reads:

```
Functors are ``functions'' from modules to modules. Functors
let you create parameterized modules and then provide other modules as parameter(s)
to get a specific implementation.  For instance, a "Set" module implementing sets
as sorted lists could be parameterized to work with any module that provides a type
and a comparison method "compare" (such as "OrderedString"):
```